Add OED estimators and accompanying tests

ae-foster opened the pull request:
This PR adds the bare bones of the new estimators for EIG and the tests for those methods. It does not change any examples (beyond deleting a couple out-dated examples). I know it weighs in at 1.5k of changes - so sorry, I did my best to keep this as small as possible without insane rewriting of the code.

Summary of changes at a code level:
 - `pyro/contrib/oed/eig.py` includes the bulk of the changes for estimators
 - `tests/contrib/oed/` includes the new tests
 - glmm is mostly compatibility changes, plus @eb8680 's Laplace guide
 - everything else (util files, etc.) is mostly bookkeeping

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pyro-ppl/pyro?pullRequest=1925) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pyro-ppl/pyro?pullRequest=1925) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pyro-ppl/pyro?pullRequest=1925) it.</sub>

eb8680 commented on pyro/contrib/oed/eig.py:
```diff
@@ -327,14 +463,180 @@ def loss_fn(design, num_particles):
+    return loss_fn
+
+
+def safe_mean_terms(terms):
```
nit: if this isn't used elsewhere and isn't meant to be called by users, rename to `_safe_mean_terms`

ae-foster commented on pyro/contrib/glmm/glmm.py:
```diff
@@ -237,11 +247,11 @@ def bayesian_linear_model(design, w_means={}, w_sqrtlambdas={}, re_group_sizes={
             k = k.expand(prediction_mean.shape)
-            transforms = [AffineTransform(loc=torch.tensor(0.), scale=k), SigmoidTransform()]
-            response_dist = dist.TransformedDistribution(base_dist, transforms)
+            response_dist = dist.CensoredSigmoidNormal(
```
Indeed- I'll revert to the old method and add the new method in a later PR

ae-foster commented on tests/contrib/oed/test_linear_models_eig.py:
```diff
@@ -0,0 +1,197 @@
+    nmc_eig, posterior_ape, marginal_eig, marginal_likelihood_eig, vnmc_eig, laplace_vi_ape, lfire_eig,
+    donsker_varadhan_eig)
+from pyro.contrib.util import rmv, rvv
+from pyro.contrib.glmm.guides import LinearModelLaplaceGuide
```
So I wrote fresh guides for the tests to make them more self-contained & readable. These guides get to be way simpler because they do not have to be prepared to all kinds of models. I could also use the `glmm/guides` versions though if you think it would be better

eb8680 replied:
I think what you have here is fine, was just curious

martinjankowiak commented:
@ae-foster thanks for the PR! haven't had a chance to look at this in detail yet but i wanted to flag a usability concern i ran into. basically all the estimators should return actual EIGs (by default, or at least optionally) and not just up to a design-independent constant. this at least wasn't the case for `posterior_ape = posterior_eig`. thoughts?

ae-foster commented:
Right now, you have `posterior_ape`, `posterior_eig_lm` (use posterior_ape and then compute the prior entropy analytically using a Gaussian assumption) and `posterior_eig_mc` (same, but compute prior entropy using Monte Carlo). If you want, we could rename as `_posterior_ape` to indicate you shouldn't use that one?

martinjankowiak commented:
hmm well wouldn't something roughly like this be cleanest?
```python
def posterior_eig(include_prior_entropy=True):
     if  include_prior_entropy:
        try:
             # analytic KL
        except:
            # MC
   _posterior_ape()
    .....```

ae-foster commented:
Yeah I can refactor it like that

ae-foster commented on pyro/contrib/oed/eig.py:
```diff
@@ -327,14 +463,180 @@ def loss_fn(design, num_particles):
+    return loss_fn
+
+
+def lfire_loss(model_marginal, model_conditional, h, observation_labels, target_labels):
```
I'm not sure you would use this directly- you can always effectively use it directly by setting `num_steps=0` in `lfire_eig`

eb8680 replied:
Makes sense, add a one-sentence docstring to that effect?

ae-foster commented on pyro/contrib/oed/eig.py:
```diff
@@ -327,14 +463,180 @@ def loss_fn(design, num_particles):
+    return loss_fn
+
+
+def vnmc_eig_loss(model, guide, observation_labels, target_labels):
```
As with `lfire`. Maybe I should make these `_` functions?

eb8680 replied:
If they aren't used outside this file, you can prefix them with `_`.  You should also add an `__all__` list to this file exporting only the things you expect to be called by users.

martinjankowiak commented:
if it helps i could write some sort of analytic KL helper along the lines of what's [here](https://github.com/pyro-ppl/pyro/blob/29bd73f476b8b9173eda3a8de71d5992888197f9/pyro/infer/trace_mean_field_elbo.py#L100)

ae-foster commented:
We should be able to re-use the old `mean_field_guide_entropy` function that we wrote before (assuming that the prior is indeed separable)

eb8680 commented on pyro/contrib/glmm/guides.py:
```diff
@@ -6,35 +6,49 @@
         # Await new version to use nn.ParamterDict
-        self._registered = nn.ParameterList(self.scale_tril.values())
+        self._registered_re = nn.ParameterList(self.regressor.values())
+        self._registered_st = nn.ParameterList(self.scale_tril.values())
```
Do you want to switch to [`torch.nn.ParameterDict`](https://pytorch.org/docs/stable/nn.html#torch.nn.ParameterDict) here now that it's available in 1.1?

ae-foster replied:
Oh yes

neerajprad commented on pyro/contrib/glmm/glmm.py:
```diff
@@ -358,7 +368,22 @@ def analytic_posterior_cov(prior_cov, x, obs_sd):
+    shape = torch.Size()
+    for s in shapes:
+        shape = torch._C._infer_size(s, shape)
+    return shape
```
There is a `torch.broadcast_tensors(*tensors)` which will directly broadcast all the tensors, so it will likely need some code modification. I remember @fritzo made a change to torch distributions to replace `torch._C._infer_size` with `torch.broadcast_tensors` for compatibility with JIT. Unless that is fixed, this will likely create issues with JIT.

ae-foster replied:
I'll rewrite to use `torch.broadcast_tensors`- this code was probably for compatibility with a now-very-outdated version to torch

martinjankowiak commented on pyro/contrib/oed/eig.py:
```diff
@@ -1,23 +1,105 @@
+]
+
+
+def laplace_vi_eig(model, design, observation_labels, target_labels, guide, loss, optim, num_steps,
```
why is this called `vi`?

ae-foster replied:
Let's change it to `laplace_eig`

martinjankowiak commented on pyro/contrib/oed/eig.py:
```diff
@@ -106,46 +217,61 @@ def naive_rainforth_eig(model, design, observation_labels, target_labels=None,
-    trace = poutine.trace(model).get_trace(expanded_design)
-    trace.compute_log_prob()
+    s = 0.
+    if N_seq > 1:
```
what's `N_seq`? this and `independent_priors` missing from docstring

ae-foster replied:
I'll document it... `N_seq > 1` actually doesn't work rn

martinjankowiak replied:
remove it?

ae-foster replied:
yes that's sensible, we'll keep it on another branch for now

martinjankowiak commented on pyro/contrib/oed/eig.py:
```diff
@@ -46,24 +128,46 @@ def vi_ape(model, design, observation_labels, target_labels,
 
     """
 
+    warnings.warn("`vi_eig` is deprecated in favour on the amortized version: `posterior_eig`.", DeprecationWarning)
```
typo: on => of

martinjankowiak commented on pyro/contrib/oed/eig.py:
```diff
@@ -231,14 +358,222 @@ def barber_agakov_ape(model, design, observation_labels, target_labels,
+    :param bool eig: Whether to compute the EIG or the average posterior entropy (APE). The EIG is given by
+                 `EIG = prior entropy - APE`. If `True`, the prior entropy will be estimated analytically,
+                 or by Monte Carlo as appropriate for the `model`. If `False` the APE is returned.
+    :param dict prior_entropy_kwargs: parameters for estimating the prior entropy.
```
namely?

martinjankowiak commented on pyro/contrib/oed/eig.py:
```diff
@@ -231,14 +358,222 @@ def barber_agakov_ape(model, design, observation_labels, target_labels,
+def vnmc_eig(model, design, observation_labels, target_labels,
+             num_samples, num_steps, guide, optim, return_history=False,
+             final_design=None, final_num_samples=None):
+    """Estimates the EIG using Variational Nested Monte Carlo (VNMC). The VNMC estimate is
```
here and elsewhere you could you some foster et al references?

ae-foster replied:
I will add a Foster et al. reference to the four estimators mentioned in our paper

martinjankowiak commented on pyro/contrib/oed/eig.py:
```diff
@@ -231,14 +358,222 @@ def barber_agakov_ape(model, design, observation_labels, target_labels,
+    where :math:`q(\\theta | y)` is the learned variational posterior approximation and
+    :math:`\\theta_n, y_n \\sim p(\\theta, y | d)`, :math:`\\theta_{mn} \\sim q(\\theta|y=y_n)`.
+
+    As :math:`N \\to \\infty` this is an upper bound on EIG. We minimise this upper bound by stochastic gradient.
```
by stochastic gradient descent?

ae-foster commented:
@martinjankowiak @eb8680 I think I've addressed all your comments now. I also removed `eig_estimation_benchmarking`: it was going to require a substantial rewrite, only to be changed again in the next PR. In later PRs we should add more user-friendly examples in `examples/contrib/oed` and maybe put numerical experiments for the paper elsewhere.
